File: ascii_fields/animations/wave_well.py
import math

from ..core import Animation, clamp, render_field


class WaveWellAnimation(Animation):
  """A quantum wave packet trapped in a 2D harmonic well. |psi|^2 is built from
  a few energy eigenstates with different phases, so the packet sloshes, spreads
  and periodically revives -- with the parabolic well walls drawn faintly."""

  thresholds = [
    (0.10, " "), (0.20, "."), (0.32, ":"), (0.44, "-"), (0.56, "="),
    (0.68, "+"), (0.80, "*"), (0.90, "#"), (1.01, "@"),
  ]

  # superposition of coherent-state lobes: (amplitude, energy, x-phase, y-phase)
  MODES = (
    (1.00, 1.0, 0.0, 0.0),
    (0.70, 2.0, 0.6, 1.1),
    (0.45, 3.0, 1.7, 0.3),
  )

  def render(self, width, height, elapsed, phase, options):
    t = elapsed * 1.3
    # packet centre traces a Lissajous path inside the well
    cx = 0.42 * math.sin(t)
    cy = 0.30 * math.sin(1.7 * t + 0.5)
    sigma = 0.30 + 0.10 * math.sin(t * 0.8)
    contrast = options.contrast
    grid = []
    for row in range(height):
      y = (row / max(1, height - 1) - 0.5) * 2.0
      line = []
      for col in range(width):
        x = (col / max(1, width - 1) - 0.5) * 2.0
        well = x * x + y * y
        # real and imaginary parts of the packet superposition
        re = 0.0
        im = 0.0
        for amp, energy, phx, phy in self.MODES:
          envelope = math.exp(-((x - cx) ** 2 + (y - cy) ** 2) / (sigma * sigma))
          ripple = math.cos(6.0 * (x * math.cos(phx) + y * math.sin(phy)) - energy * t)
          rippl2 = math.sin(6.0 * (x * math.cos(phx) + y * math.sin(phy)) - energy * t)
          re += amp * envelope * ripple
          im += amp * envelope * rippl2
        prob = re * re + im * im
        wall = 0.06 * math.exp(-((math.sqrt(well) - 0.95) ** 2) / 0.01)
        level = clamp(prob * 0.5 * contrast + wall)
        line.append(level)
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/wave_well.rs
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "ocean" };
const TH: &[(f64, char)] = &[
  (0.10, ' '),
  (0.20, '.'),
  (0.32, ':'),
  (0.44, '-'),
  (0.56, '='),
  (0.68, '+'),
  (0.80, '*'),
  (0.90, '#'),
  (1.01, '@'),
];
// (amplitude, energy, x-phase, y-phase)
const MODES: &[(f64, f64, f64, f64)] = &[(1.00, 1.0, 0.0, 0.0), (0.70, 2.0, 0.6, 1.1), (0.45, 3.0, 1.7, 0.3)];

pub struct WaveWell;

impl Animation for WaveWell {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width;
    let h = ctx.height;
    let t = ctx.elapsed * 1.3;
    let cx = 0.42 * t.sin();
    let cy = 0.30 * (1.7 * t + 0.5).sin();
    let sigma = 0.30 + 0.10 * (t * 0.8).sin();
    let contrast = ctx.options.contrast;
    let mut grid = vec![0.0_f64; w * h];
    let dw = (w.saturating_sub(1)).max(1) as f64;
    let dh = (h.saturating_sub(1)).max(1) as f64;
    for row in 0..h {
      let y = (row as f64 / dh - 0.5) * 2.0;
      let base = row * w;
      for col in 0..w {
        let x = (col as f64 / dw - 0.5) * 2.0;
        let well = x * x + y * y;
        let mut re = 0.0_f64;
        let mut im = 0.0_f64;
        for &(amp, energy, phx, phy) in MODES {
          let envelope = (-((x - cx).powi(2) + (y - cy).powi(2)) / (sigma * sigma)).exp();
          let ripple = (6.0 * (x * phx.cos() + y * phy.sin()) - energy * t).cos();
          let rippl2 = (6.0 * (x * phx.cos() + y * phy.sin()) - energy * t).sin();
          re += amp * envelope * ripple;
          im += amp * envelope * rippl2;
        }
        let prob = re * re + im * im;
        let wall = 0.06 * (-((well.sqrt() - 0.95).powi(2)) / 0.01).exp();
        grid[base + col] = clamp(prob * 0.5 * contrast + wall);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
